### PR TITLE
Update compiler-rt to llvm-12

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

**Comment by sbc100** [comment]
To see the diff from upstream you can use this like:
https://github.com/llvm/llvm-project/compare/llvmorg-12.0.0...emscripten-core:emscripten-libs-12.0.0

File: src/library_pthread.js [hunk 1]
```diff
@@ -1043,9 +1043,13 @@ var LibraryPThread = {
...
   },
 
-  // C11 threads function.
+  // C11 thread version.
   // TODO: remove this in favor or compiling musl/src/thread/pthread_detach.c
+#if USE_LSAN
+  thrd_detach: 'emscripten_builtin_pthread_detach',
+#else
   thrd_detach: 'pthread_detach',
+#endif
```
Comment (kripken) [inline src/library_pthread.js:1052]: this could be on a single line like the `{{{ }}}` stuff above.

**Review by kripken — APPROVED** [review]
Looking at the diff to upstream:

- Is the diff in `ubsan_checks.inc` necessary?
- The diff in `ubsan_handlers.cpp` looks odd - no ifdefing for emscripten, just a pure change to upstream?

File: src/library_pthread.js (hunk 1 above)
Comment (sbc100) [reply src/library_pthread.js:1052]: I couldn't make it work when the string is on the right of the ":" for some reason.  It was giving me JS compiler errors.   Plus I think this syntax is more readable in the end.

**Comment by sbc100** [comment]
> Looking at the diff to upstream:
> 
> * Is the diff in `ubsan_checks.inc` necessary?

These are part of the existing 10.0.0 changes.  I don't really know why they were removed, but they were.   Maybe something to do with the way emscripten handles NULL vs native platforms.

See the existing 10.0.0 changes:
https://github.com/llvm/llvm-project/compare/llvmorg-10.0.0...emscripten-core:emscripten-libs-10.0.0#diff-8ced148882cd9876de0c3b050db5402f972d0093e2762a750177890510718c30

> * The diff in `ubsan_handlers.cpp` looks odd - no ifdefing for emscripten, just a pure change to upstream?

Again, these are just carred over from the current changes: 

https://github.com/llvm/llvm-project/compare/llvmorg-10.0.0...emscripten-core:emscripten-libs-10.0.0#diff-f899d1244fc94738aa307e798a3d7bd442539e5c97e4a93bae29dbae992195c8

We can look at cleaning them or and/removing as we look into upstreaming perhaps?

File: src/library_pthread.js (hunk 1 above)
Comment (kripken) [reply src/library_pthread.js:1052]: But it works just several lines above in the diff? I don't understand. It's just adding a prefix to a string, isn't it?

Comment (sbc100) [reply src/library_pthread.js:1052]: Somehow it works in the key position but not in the value position.   Oh maybe its because the keys are not in quotes..  Let me try again.

**Comment by aheejin** [comment]
It looks several asan/lsan tests started to break after this. Starting commit: https://ci.chromium.org/ui/p/emscripten-releases/builders/ci/linux-test-suites/b8846238787694172272/overview

**Comment by sbc100** [comment]
Ah yes dropped the ball on that..   thanks for reminding me.
